### PR TITLE
VITIS-10045 - add prototype

### PR DESCRIPTION
**Pull request by dezhiAmd**
Problem solved in this commit:

- Support parsing elf file generated from an instruction file and control-code file. (Refer to [Instruction buffer and control buffer patching ](https://gitenterprise.xilinx.com/XRT/ERT4/pull/91)

Change:

-  Add new patching method for IPU driver. Refer to [xrt_module.cpp change](https://github.com/larry9523/XRT/commit/d7a5941dc378a710cf7437ad6d899dee7e481d1b)

Signed-off-by: Dezhi Liao [dezhi.liao@amd.com](mailto:dezhi.liao@amd.com)

### CODE REVIEW

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 1]
```diff
@@ -32,593 +33,1058 @@
...
+    // 0 if no padding is required.   The page size should be
+    // embedded as ELF metadata in the future.
+    static constexpr size_t column_page_size = AIE_COLUMN_PAGE_SIZE;
+
+    // struct instr_buf - represent instruction buffer
+    // TODO place holder in case we need special handling
+    // TODO merge with ctrlcode?
+    struct instr_buf
+    {
+        std::vector<uint8_t> m_data;
```
Comment (stsoe) [inline src/runtime_src/core/common/api/xrt_module.cpp:46]: Please use existing code style within the file.   For api/ files the convention is

- no tabs, 2 spaces scope indent
- new line after return type for all functions to line up as much as possible to the left, e.g.

```
type 
function(arg1, arg)
{
   body;
}
```

I attempted to edit a few changes below as an example. 
Else everything looks good.  Thank you!

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 2]
```diff
@@ -32,593 +33,1058 @@
...
+
+        void
+            append_section_data(ELFIO::section* sec)
```
Comment (stsoe) [inline src/runtime_src/core/common/api/xrt_module.cpp:49]:
```suggestion
    void
    append_section_data(ELFIO::section* sec)
```

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 3]
```diff
@@ -32,593 +33,1058 @@
...
+    //
+    // Manage patching of a symbol in the control code.  The symbol
+    // type is used to determine the patching method.
+    //
+    // The patchers is created with offsets into a buffer object.
+    // The base address of the buffer object is passed
+    // in as a parameter to patch().
+    struct patchers
+    {
+        enum class symbol_type {
```
Comment (stsoe) [inline src/runtime_src/core/common/api/xrt_module.cpp:157]:
```suggestion
      enum class symbol_type {
```

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 4]
```diff
@@ -32,593 +33,1058 @@
...
+        // The base address is passed in as a parameter to patch()
+        std::vector<uint64_t> m_ctrlcode_offset;
+
+        patchers(symbol_type type, std::string name, std::vector<uint64_t> ctrlcode_offset)
+            : m_symbol_type(type)
+            , m_buf_name(name)
+            , m_ctrlcode_offset(ctrlcode_offset)
+        {}
+
+        void patch32(uint32_t* bd_data_ptr, uint64_t patch)
```
Comment (stsoe) [inline src/runtime_src/core/common/api/xrt_module.cpp:178]:
```suggestion
      void 
      patch32(uint32_t* bd_data_ptr, uint64_t patch)
```

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 5]
```diff
@@ -32,593 +33,1058 @@
...
+        // structures that manages the control code data for each column and
+        // page, then create ctrlcode objects from the data.
+        static std::vector<ctrlcode>
+            initialize_column_ctrlcode(const ELFIO::elfio& elf)
+        {
+            auto os_abi = elf.get_os_abi();
+            if (os_abi != ELFIO::ELFOSABI_ARM) {
+                XRT_PRINTF("module_elf::initialize_column_ctrlcode(), not AIE2S, skip...\n");
+                return {};
+            }
```
Comment (stsoe) [inline src/runtime_src/core/common/api/xrt_module.cpp:541]:
```suggestion
      static std::vector<ctrlcode>
      initialize_column_ctrlcode(const ELFIO::elfio& elf)
      {
        auto os_abi = elf.get_os_abi();
        if (os_abi != ELFIO::ELFOSABI_ARM) {
          XRT_PRINTF("module_elf::initialize_column_ctrlcode(), not AIE2S, skip...\n");
          return {};
       }
```

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 6]
```diff
@@ -32,593 +33,1058 @@
...
+
+            // Elf sections for a single page
+            struct column_page {
+                ELFIO::section* ctrltext = nullptr;
+                ELFIO::section* ctrldata = nullptr;
+            };
```
Comment (stsoe) [inline src/runtime_src/core/common/api/xrt_module.cpp:547]:
```suggestion
          struct column_page {
            ELFIO::section* ctrltext = nullptr;
            ELFIO::section* ctrldata = nullptr;
          };
```

**Review by stsoe — CHANGES_REQUESTED** [review]
Some code style changes to stay consistent with all files under core/common/api.
I apologize for the non standard conventions but I prefer as much lined up to the left as possible.

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 7]
```diff
@@ -30,13 +31,70 @@
...
 
 // Control code is padded to page size, where page size is
 // 0 if no padding is required.   The page size should be
 // embedded as ELF metadata in the future.
 static constexpr size_t column_page_size = AIE_COLUMN_PAGE_SIZE;
 
+// struct instr_buf - represent instruction buffer
+// TODO place holder in case we need special handling
+// TODO merge with ctrlcode?
+struct instr_buf
```
Comment (stsoe) [inline src/runtime_src/core/common/api/xrt_module.cpp:45]: struct instr_buf is identical to struct control_packet, why two structs?
also looks kind of similar to struct ctrlcode

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 8]
```diff
@@ -431,12 +753,33 @@ class module_elf : public module_impl
...
 public:
   module_elf(xrt::elf elf)
-    : module_impl{elf.get_cfg_uuid()}
+    : module_impl{ elf.get_cfg_uuid() }
     , m_elf(std::move(elf))
-    , m_ctrlcodes{initialize_column_ctrlcode(xrt_core::elf_int::get_elfio(m_elf))}
-    , m_arg2patcher{initialize_arg_patchers(xrt_core::elf_int::get_elfio(m_elf), m_ctrlcodes)}
+    , m_os_abi{ xrt_core::elf_int::get_elfio(m_elf).get_os_abi() }
+    , m_ctrlcodes{ initialize_column_ctrlcode(xrt_core::elf_int::get_elfio(m_elf)) }
+    , m_arg2patcher{ initialize_arg_patchers(xrt_core::elf_int::get_elfio(m_elf), m_ctrlcodes) }
```
Comment (stsoe) [inline src/runtime_src/core/common/api/xrt_module.cpp:779]: What is the different between the two m_arg2patcher(s) ?  Do we need both?
E.g. do we need the old original m_ctrlcodes and m_arg2patchers?

**Review by stsoe — COMMENTED** [review]
Thanks for reformatting.  Looks clean.
Few questions, I am trying to understand where we going with this and what the role is of the legacy original code?

File: src/runtime_src/core/common/api/xrt_module.cpp (hunk 7 above)
Comment (dezhiAmd) [reply src/runtime_src/core/common/api/xrt_module.cpp]: Keep only ctrlcode.

File: src/runtime_src/core/common/api/xrt_module.cpp (hunk 8 above)
Comment (dezhiAmd) [reply src/runtime_src/core/common/api/xrt_module.cpp:723]: m_arg2patcher is for Turin
m_arg2patchers is for IPU.
A follow-up PR will merge them after testing on Turin

**Review by stsoe — APPROVED** [review]
Looks good to me.  Thank you.
Minor nitpick (1) no first level indent within namespace, (2) curly brace same line as namespace.   Neither matters much to me, but in this directory and this file it is convention for all files.

**Comment by dezhiAmd** [comment]
retest this please.

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 9]
```diff
@@ -590,25 +981,53 @@ class module_sram : public module_impl
...
   sync_if_dirty() override
   {
+#if 0
     if (m_patched_args.size() != m_parent->number_of_arg_patchers()) {
       auto fmt = boost::format("ctrlcode requires %d patched arguments, but only %d are patched")
         % m_parent->number_of_arg_patchers() % m_patched_args.size();
-      throw std::runtime_error{fmt.str()};
+      throw std::runtime_error{ fmt.str() };
     }
+#endif
```
Comment (larry9523) [inline src/runtime_src/core/common/api/xrt_module.cpp:990]: We should get rid of this "if 0". It was commented out for the quick prototyping. Need to understand what it is doing and enable it for all use cases.

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 10]
```diff
@@ -590,25 +981,53 @@ class module_sram : public module_impl
...
+      m_buffer.sync(XCL_BO_SYNC_BO_TO_DEVICE);
+
+#if 0
+    // debug code to save patched buffer onto disk
+    uint32_t* p = reinterpret_cast<uint32_t*>(m_ctrlpkt_buf.map<char*>());
+    auto buff = reinterpret_cast<char*>(p);
+    std::ofstream outdata("./xrt_patched", std::ios::out | std::ios::binary);
+    outdata.write(buff, 4096);
+    outdata.close();
+#endif
```
Comment (larry9523) [inline src/runtime_src/core/common/api/xrt_module.cpp:1010]: Should remove this code

File: src/runtime_src/core/common/api/xrt_module.cpp (hunk 9 above)
Comment (dezhiAmd) [reply src/runtime_src/core/common/api/xrt_module.cpp:990]: Agree. But this needs to be done in a follow up pull request where we can merge m_arg2patcher (used by TurinAI) with m_arg2patchers (used by IPU)

File: src/runtime_src/core/common/api/xrt_module.cpp (hunk 10 above)
Comment (dezhiAmd) [reply src/runtime_src/core/common/api/xrt_module.cpp]: Fixed

File: src/runtime_src/core/common/api/xrt_module.cpp (hunk 9 above)
Comment (dezhiAmd) [reply src/runtime_src/core/common/api/xrt_module.cpp]: Fixed

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 11]
```diff
@@ -551,8 +933,13 @@ class module_sram : public module_impl
...
       return acc + ctrlcode.size();
-    });
-    m_buffer = xrt::bo{m_hwctx, sz, xrt::bo::flags::cacheable, 1 /* fix me */};
+      });
+    if (sz == 0) {
+      std::cout << "instruction buf is empty" << std::endl;
+      return;
+    }
+
+    m_buffer = xrt::bo{ m_hwctx, sz, xrt::bo::flags::cacheable, 1 /* fix me */ };
```
Comment (dezhiAmd) [inline src/runtime_src/core/common/api/xrt_module.cpp:942]: What needs to be fixed in this line?

Comment (stsoe) [reply src/runtime_src/core/common/api/xrt_module.cpp:942]: The API requiring `1` for memory bank is awkward.   The `xrt:;bo::cacheable` flag is a bad hack also.

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 12]
```diff
@@ -301,15 +484,24 @@ class module_elf : public module_impl
   static std::vector<ctrlcode>
   initialize_column_ctrlcode(const ELFIO::elfio& elf)
   {
+    auto os_abi = elf.get_os_abi();
+    if (os_abi != ELFIO::ELFOSABI_ARM)
```
Comment (dezhiAmd) [inline src/runtime_src/core/common/api/xrt_module.cpp:488]: This breaks

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 13]
```diff
@@ -358,9 +550,83 @@ class module_elf : public module_impl
...
+      }
+    }
+
+    return arg2patchers;
+  }
+
   static std::map<std::string, patcher>
   initialize_arg_patchers(const ELFIO::elfio& elf, const std::vector<ctrlcode>& ctrlcodes)
   {
+    auto os_abi = elf.get_os_abi();
```
Comment (dezhiAmd) [inline src/runtime_src/core/common/api/xrt_module.cpp:624]: Not needed

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 14]
```diff
@@ -285,13 +412,69 @@ class module_elf : public module_impl
...
+    return instrbuf;
+  }
+
+  // Extract control-packet buffer from ELF sections without assuming anything
+  // about order of sections in the ELF file.
+  static control_packet
+  initialize_ctrl_packet(const ELFIO::elfio& elf)
+  {
+    auto os_abi = elf.get_os_abi();
+    if (os_abi != 66) {
```
Comment (larry9523) [inline src/runtime_src/core/common/api/xrt_module.cpp]: same comment

**Review by larry9523 — COMMENTED** [review]
Looks good. Let's test on Turin and merge it after we prove this won't break anything there.

**Comment by manikandan-xilinx** [comment]
retest this please.

**Comment by dezhiAmd** [comment]
Test result are successful both on IPU and Turin Simnow. Refer to the screen shot below.
Screen shot of test result on IPU:

image


Screen shot of test result on Turin Simnow:

image

**Review by larry9523 — APPROVED** [review]
Looking forward to the next PR for merging the two ELF patching supports

**Comment by dezhiAmd** [comment]
> Looking forward to the next PR for merging the two ELF patching supports

Refer to this PR [PR7807](https://github.com/Xilinx/XRT/pull/7807/)

**Pull request merged**